common/custom_claims.c

```diff
@@ -0,0 +1,222 @@
+}
+
+static oe_result_t _free_claims(oe_claim_t* claims, size_t claims_length)
+{
+    if (!claims)
+        return OE_OK;
+
+    for (size_t i = 0; i < claims_length; i++)
+        _free_claim(&claims[i]);
+    oe_free(claims);
+    return OE_OK;
+}
+
+static oe_result_t _add_claim(
+    oe_claim_t* claim,
+    const void* name,
+    size_t name_size, // Must cover the '\0' at end of string
+    const void* value,
+    size_t value_size)
+{
+    if (*((uint8_t*)name + name_size - 1) != '\0')
+        return OE_CONSTRAINT_FAILED;
+
+    claim->name = (char*)oe_malloc(name_size);
+    if (claim->name == NULL)
+        return OE_OUT_OF_MEMORY;
+    memcpy(claim->name, name, name_size);
+
+    claim->value = (uint8_t*)oe_malloc(value_size);
+    if (claim->value == NULL)
+    {
+        oe_free(claim->name);
+        claim->name = NULL;
+        return OE_OUT_OF_MEMORY;
+    }
+    memcpy(claim->value, value, value_size);
+    claim->value_size = value_size;
+
+    return OE_OK;
+}
+
+static oe_result_t _fill_with_custom_claims(
+    const uint8_t* claims_buffer,
+    size_t claims_buffer_size,
+    oe_claim_t* claims)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_custom_claims_header_t* header =
+        (oe_custom_claims_header_t*)claims_buffer;
+    size_t claims_index = 0;
+
+    claims_buffer += sizeof(*header);
+    claims_buffer_size -= sizeof(*header);
+    for (uint64_t i = 0; i < header->num_claims; i++)
+    {
+        oe_custom_claims_entry_t* entry =
+            (oe_custom_claims_entry_t*)claims_buffer;
+        uint64_t size = 0;
+
+        // Sanity check sizes.
+        if (claims_buffer_size < sizeof(*entry))
+            OE_RAISE(OE_CONSTRAINT_FAILED);
+
+        OE_CHECK(oe_safe_add_u64(sizeof(*entry), entry->name_size, &size));
+        OE_CHECK(oe_safe_add_u64(size, entry->value_size, &size));
+
+        if (claims_buffer_size < size)
+            OE_RAISE(OE_CONSTRAINT_FAILED);
+
+        // Add the claim.
+        OE_CHECK(_add_claim(
+            &claims[claims_index++],
+            entry->name,
+            entry->name_size,
+            entry->name + entry->name_size,
+            entry->value_size));
+
+        // Go to next entry.
+        claims_buffer += size;
+        claims_buffer_size -= size;
+    }
+
+    result = OE_OK;
+
+done:
+    if (result != OE_OK)
+    {
+        for (size_t i = 0; i < claims_index; i++)
+            _free_claim(&claims[i]);
+    }
+    return result;
+}
+
+oe_result_t oe_deserialize_custom_claims(
+    const uint8_t* claims_buffer,
+    size_t claims_buffer_size,
+    oe_claim_t** claims_out,
+    size_t* claims_length_out)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_custom_claims_header_t* claims_header = NULL;
+    oe_claim_t* claims = NULL;
+    uint64_t claims_length = 0;
+    uint64_t claims_size = 0;
+
+    claims_header = (oe_custom_claims_header_t*)(claims_buffer);
+    if (claims_header == NULL)
+    {
+        OE_RAISE(OE_INVALID_PARAMETER);
+    }
+
```

Comment:
Should check that claim buffer size is at least sufficient for the header, and the version is expected.

Reply:
add the checks